File: apps/android/core/player/src/main/kotlin/net/koalastuff/koalacast/core/player/Spectrum.kt
package net.koalastuff.koalacast.core.player

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * The frequency half of the visualiser signal.
 *
 * A single loudness envelope can only ever drive one shape that swells and
 * shrinks. Anything that looks like the bar meter people expect from a media
 * player — bars that stand still horizontally and only change height, each one
 * answering for a different part of the sound — needs the spectrum, so this
 * takes the same decoded PCM [AmplitudeTap] already sees and runs an FFT over it.
 *
 * Still no `RECORD_AUDIO`: this is our own audio, read on its way to the sink.
 */

/**
 * 2048 points is ~46 ms of audio at 44.1 kHz and ~21 Hz per bin.
 *
 * This is the whole trade. Shorter windows react faster but resolve worse, and
 * at 512 points the bins are 86 Hz apart — wider than the bottom third of a
 * log-spaced display, so those bands collapse onto each other and eighteen bars
 * end up drawing the same number. Longer windows fix that and start showing the
 * syllable rather than the consonant. 2048 leaves the bottom few bands sharing
 * bins, which is invisible, and keeps the display within a frame or two of the
 * sound. The frames overlap by three quarters, so a new spectrum still arrives
 * about every 12 ms.
 */
internal const val SPECTRUM_FFT_SIZE = 2048

/**
 * How many bars a spectrum style draws. Enough to fill a phone's width at a
 * readable bar width without turning each bar into a hairline.
 */
const val SPECTRUM_BANDS = 48

/** Below this is rumble, above it is hiss; neither says anything about speech. */
private const val SPECTRUM_MIN_HZ = 60.0
private const val SPECTRUM_MAX_HZ = 12_000.0

/**
 * The visible window, in dBFS. Quieter than the floor reads as an empty band.
 *
 * The ceiling was -14 dB, which ordinary mastered speech passes for most of a
 * sentence: nearly every band pinned at full height, the shape stopped having
 * any range and the display read as a solid bar with a fringe. -4 leaves the
 * headroom that mastering actually uses, so a loud passage has somewhere to go.
 */
private const val SPECTRUM_FLOOR_DB = -78f
private const val SPECTRUM_CEILING_DB = -4f

/**
 * Recorded speech rolls off with frequency; without this the right half is dead.
 * Kept modest — at 1.6 the top bands were lifted by 2.6x and clipped on their
 * own, which is the same saturation from the other end.
 */
private const val SPECTRUM_TILT = 0.8f

/**
 * In-place radix-2 Cooley–Tukey FFT over [re] and [im], which must be the same
 * power-of-two length.
 *
 * Hand-written rather than pulled in from a DSP library because it runs on the
 * audio thread: it must allocate nothing, and a dependency that allocates one
 * `Complex[]` per call would drop buffers on cheap devices. Only the real input
 * case is used here, so the imaginary half arrives zeroed.
 */
internal fun fftInPlace(re: FloatArray, im: FloatArray) {
    val n = re.size

    // Bit-reversal permutation, computed by incrementing a reversed counter
    // rather than by reversing each index, so there is no per-element loop.
    var j = 0
    for (i in 1 until n) {
        var bit = n shr 1
        while (j and bit != 0) {
            j = j xor bit
            bit = bit shr 1
        }
        j = j or bit
        if (i < j) {
            val tr = re[i]; re[i] = re[j]; re[j] = tr
            val ti = im[i]; im[i] = im[j]; im[j] = ti
        }
    }

    var length = 2
    while (length <= n) {
        val angle = -2.0 * PI / length
        val wRealStep = cos(angle).toFloat()
        val wImagStep = sin(angle).toFloat()
        var start = 0
        while (start < n) {
            var wReal = 1f
            var wImag = 0f
            for (offset in 0 until length / 2) {
                val a = start + offset
                val b = a + length / 2
                val tr = wReal * re[b] - wImag * im[b]
                val ti = wReal * im[b] + wImag * re[b]
                re[b] = re[a] - tr
                im[b] = im[a] - ti
                re[a] += tr
                im[a] += ti
                val nextReal = wReal * wRealStep - wImag * wImagStep
                wImag = wReal * wImagStep + wImag * wRealStep
                wReal = nextReal
            }
            start += length
        }
        length = length shl 1
    }
}

/**
 * The FFT bin index each band starts at, plus a final entry for the end of the
 * last band, so a band's bins are `edges[i] until edges[i + 1]`.
 *
 * Log-spaced, because linear spacing puts nine tenths of the display above 4 kHz
 * where speech has almost nothing, and the result is a row of bars in which only
 * the leftmost two ever move. Each band is guaranteed at least one bin so no
 * band renders as a permanent gap.
 */
internal fun spectrumBandEdges(
    sampleRateHz: Int,
    fftSize: Int = SPECTRUM_FFT_SIZE,
    bands: Int = SPECTRUM_BANDS,
): IntArray {
    val usableBins = fftSize / 2
    val nyquist = sampleRateHz.coerceAtLeast(1) / 2.0
    val logMin = ln(SPECTRUM_MIN_HZ)
    val logMax = ln(SPECTRUM_MAX_HZ.coerceAtMost(nyquist).coerceAtLeast(SPECTRUM_MIN_HZ * 2))
    val edges = IntArray(bands + 1)
    for (band in 0..bands) {
        val hz = exp(logMin + (logMax - logMin) * band / bands)
        edges[band] = ((hz / nyquist) * usableBins).toInt().coerceIn(0, usableBins)
    }
    // The bottom bands are narrower than one bin at this FFT size, so they would
    // otherwise collapse onto each other and render as a permanent gap. Pushing
    // each edge past the one before it costs a little accuracy at the very bottom
    // and guarantees every bar has something to report. Safe without an upper
    // clamp only because there are far more usable bins than bands.
    require(bands < usableBins) { "spectrum needs more FFT bins than bands" }
    for (band in 1..bands) {
        if (edges[band] <= edges[band - 1]) edges[band] = edges[band - 1] + 1
    }
    return edges
}

/** A Hann window of [size]. Rectangular windows smear a tone across five bands. */
internal fun hannWindow(size: Int): FloatArray =
    FloatArray(size) { 0.5f - 0.5f * cos(2.0 * PI * it / (size - 1)).toFloat() }

/**
 * Reduces one FFT result to [out] band energies, 0..1, low frequencies first.
 *
 * Peak within a band rather than mean: averaging across a band that spans several
 * kHz buries every transient, and transients are the part a listener recognises.
 */
internal fun reduceToBands(
    re: FloatArray,
    im: FloatArray,
    edges: IntArray,
    out: FloatArray,
) {
    val scale = 2f / re.size
    for (band in out.indices) {
        var peak = 0f
        for (bin in edges[band] until edges[band + 1]) {
            val magnitude = sqrt(re[bin] * re[bin] + im[bin] * im[bin]) * scale
            if (magnitude > peak) peak = magnitude
        }
        val db = if (peak > 1e-7f) 20f * log10(peak) else SPECTRUM_FLOOR_DB
        val normalised = (db - SPECTRUM_FLOOR_DB) / (SPECTRUM_CEILING_DB - SPECTRUM_FLOOR_DB)
        val tilt = 1f + SPECTRUM_TILT * band / (out.size - 1).coerceAtLeast(1)
        out[band] = (normalised * tilt).coerceIn(0f, 1f)
    }
}

/**
 * Where a loud passage should land, and how far the gain may reach for it.
 *
 * Normalising against full scale is why a quietly mastered episode drew a flat
 * display while a loud one drew a lively one. A music player's spectrum looks
 * alive at every volume because it is normalised against what it has been
 * hearing. Podcast audio makes that more pronounced still: levelling between
 * shows is far less consistent than in mastered music.
 *
 * Deliberately identical to the web client's `spectrum.ts`, so the two displays
 * answer alike.
 */
private const val AGC_TARGET = 0.82f
private const val AGC_MAX_GAIN = 3f

/** Below this the frame is silence or room tone, and lifting it only draws noise. */
internal const val AGC_SILENCE = 0.06f

/** Rises quickly so a transient pulls the gain down at once; falls slowly. */
private const val AGC_ATTACK = 0.35f
private const val AGC_RELEASE = 0.015f

/** The loudest band this display has been seeing, smoothed. */
internal class AutoGain {
    var reference: Float = 0f
        private set

    fun reset() {
        reference = 0f
    }

    /**
     * Scales [bands] in place so the display uses its height at any input level,
     * and returns the gain applied.
     *
     * Silence is left alone on purpose: an empty display during a pause is
     * correct, and amplifying room tone into a full-height wall is not.
     */
    fun apply(bands: FloatArray): Float {
        var frontRunner = 0f
        for (value in bands) if (value > frontRunner) frontRunner = value
        val coefficient = if (frontRunner > reference) AGC_ATTACK else AGC_RELEASE
        reference += (frontRunner - reference) * coefficient

        if (reference < AGC_SILENCE) return 1f
        val gain = (AGC_TARGET / reference).coerceIn(1f, AGC_MAX_GAIN)
        if (gain == 1f) return 1f
        for (index in bands.indices) bands[index] = (bands[index] * gain).coerceAtMost(1f)
        return gain
    }
}

File: apps/web/src/lib/audio/spectrum.ts
// The frequency half of the visualiser signal.
//
// Kept apart from the audio graph so the arithmetic can be tested without a
// browser: an AnalyserNode cannot be driven from a unit test, and every failure
// this file exists to prevent is a failure of the mapping rather than of the
// audio plumbing.
//
// The Android client reaches the same numbers through its own FFT (see
// core/player/Spectrum.kt). The constants below are deliberately identical to
// its, because the two displays are the same product and drifting tunings are
// how one of them ends up looking broken while the other does not — which is
// exactly what happened here.

/** Below this is rumble, above it is hiss; neither says anything about speech. */
export const SPECTRUM_MIN_HZ = 60;
export const SPECTRUM_MAX_HZ = 12_000;

/**
 * The visible window, in dBFS.
 *
 * The AnalyserNode's defaults are -100 and -30, and this file used to leave
 * them alone. A -30 ceiling is below the level ordinary mastered speech spends
 * most of a sentence at, so band after band sat pinned at 255 — and a bar that
 * is clipped does not move. Half the display looking frozen was that.
 */
export const SPECTRUM_FLOOR_DB = -78;
export const SPECTRUM_CEILING_DB = -4;

/**
 * Recorded speech rolls off with frequency; without this the right half is dead.
 * Modest on purpose: at 1.6 the top bands are lifted by 2.6x and clip on their
 * own, which is the same frozen display arriving from the other end.
 */
export const SPECTRUM_TILT = 0.8;

/**
 * The FFT bin index each band starts at, plus a final entry for the end of the
 * last band, so a band's bins are `edges[i]` until `edges[i + 1]`.
 *
 * Log-spaced: linear spacing puts nine tenths of the display above 4 kHz where
 * speech has almost nothing, leaving a row of bars in which only the leftmost
 * two ever move. Every band is guaranteed at least one bin, so no bar renders
 * as a permanent gap.
 */
export function spectrumBandEdges(
	sampleRateHz: number,
	binCount: number,
	bands: number
): Int32Array {
	if (bands >= binCount) throw new Error('spectrum needs more FFT bins than bands');
	const nyquist = Math.max(1, sampleRateHz) / 2;
	const logMin = Math.log(SPECTRUM_MIN_HZ);
	const logMax = Math.log(
		Math.max(SPECTRUM_MIN_HZ * 2, Math.min(SPECTRUM_MAX_HZ, nyquist))
	);
	const edges = new Int32Array(bands + 1);
	for (let band = 0; band <= bands; band++) {
		const hz = Math.exp(logMin + ((logMax - logMin) * band) / bands);
		edges[band] = Math.min(binCount, Math.max(0, Math.floor((hz / nyquist) * binCount)));
	}
	// The bottom bands are narrower than one bin at any practical FFT size, so
	// without this they collapse onto each other and a stretch of bars all draw
	// the same number. Safe without an upper clamp only because there are far
	// more bins than bands, which the guard above enforces.
	for (let band = 1; band <= bands; band++) {
		if (edges[band] <= edges[band - 1]) edges[band] = edges[band - 1] + 1;
	}
	return edges;
}

/**
 * Reduces one byte-magnitude spectrum to per-band energies, 0..1.
 *
 * `data` is what `AnalyserNode.getByteFrequencyData` produces, which is already
 * the dB window mapped onto 0..255 — provided the window was configured; see
 * SPECTRUM_FLOOR_DB. Peak within a band rather than mean, because averaging
 * across a band spanning several kHz buries every transient, and transients are
 * the part a listener recognises.
 */
export function reduceToBands(data: Uint8Array, edges: Int32Array, out: Float32Array): void {
	for (let band = 0; band < out.length; band++) {
		let peak = 0;
		const end = Math.min(data.length, edges[band + 1]);
		for (let bin = edges[band]; bin < end; bin++) {
			if (data[bin] > peak) peak = data[bin];
		}
		const tilt = 1 + (SPECTRUM_TILT * band) / Math.max(1, out.length - 1);
		out[band] = Math.min(1, (peak / 255) * tilt);
	}
}

/**
 * Where a loud passage should land, and how far the gain may reach for it.
 *
 * This is the part neither client had, and the reason a quiet episode drew a
 * flat display while a loud one drew a lively one. A music player's spectrum
 * looks alive at every volume because it is normalised against what it has been
 * hearing, not against full scale. Podcast audio makes it more pronounced still:
 * levelling between shows is far less consistent than in mastered music.
 */
export const AGC_TARGET = 0.82;
export const AGC_MAX_GAIN = 3;
/** Below this the frame is silence or room tone, and lifting it only draws noise. */
export const AGC_SILENCE = 0.06;
/** Rises quickly so a transient pulls the gain down at once; falls slowly. */
export const AGC_ATTACK = 0.35;
export const AGC_RELEASE = 0.015;

export interface AutoGainState {
	/** The loudest band this display has been seeing, smoothed. */
	reference: number;
}

export function createAutoGainState(): AutoGainState {
	return { reference: 0 };
}

/**
 * Scales `bands` in place so the display uses its height at any input level.
 *
 * Returns the gain applied, for tests and for callers that want to reason about
 * it. Silence is deliberately left alone: an empty display during a pause is
 * correct, and amplifying room tone into a full-height wall is not.
 */
export function applyAutoGain(bands: Float32Array, state: AutoGainState): number {
	let frontRunner = 0;
	for (let band = 0; band < bands.length; band++) {
		if (bands[band] > frontRunner) frontRunner = bands[band];
	}
	const coefficient = frontRunner > state.reference ? AGC_ATTACK : AGC_RELEASE;
	state.reference += (frontRunner - state.reference) * coefficient;

	if (state.reference < AGC_SILENCE) return 1;
	const gain = Math.min(AGC_MAX_GAIN, Math.max(1, AGC_TARGET / state.reference));
	if (gain === 1) return 1;
	for (let band = 0; band < bands.length; band++) {
		bands[band] = Math.min(1, bands[band] * gain);
	}
	return gain;
}
